spec/spec_helper.rb

```diff
@@ -42,3 +42,5 @@ def select_io_pair_by_name(category, name)
 def select_input_by_name(category, name)
   select_io_pair_by_name(category, name)[0]
 end
+
+RSpec::Expectations.configuration.on_potential_false_positives = :nothing
```

Comment:
I don't think that it should be disabled for any spec. 
for further discussion: https://github.com/rspec/rspec-expectations/issues/231

Reply:
This is disabled to avoid having numerous false-positive warnings about risking false positives. 

The background for this specific rule is that without it, our tests are often sprinkled with warning messages: 

```
JsonApi::Parameters::Validator
  initializer
    ensures @payload has keys deeply stringified
    validations
      loads JsonApi schema
      with prevalidation enforced
        raises validation errors
WARNING: Using `expect { }.not_to raise_error(SpecificErrorClass)` risks false positives, since literally any other error would cause the expectation to pass, including those raised by Ruby (e.g. NoMethodError, NameError and ArgumentError), meaning the code you are intending to test may not even get reached. Instead consider using `expect { }.not_to raise_error` or `expect { }.to raise_error(DifferentSpecificErrorClass)`. This message can be suppressed by setting: `RSpec::Expectations.configuration.on_potential_false_positives = :nothing`. Called from /home/marahin/Code/jsonapi_parameters/spec/lib/jsonapi_parameters/validator_spec.rb:35:in `block (5 levels) in <top (required)>'.
        does not raise TranslatorError
        does not call formed_parameters
      suppression enabled
WARNING: Using `expect { }.not_to raise_error(SpecificErrorClass)` risks false positives, since literally any other error would cause the expectation to pass, including those raised by Ruby (e.g. NoMethodError, NameError and ArgumentError), meaning the code you are intending to test may not even get reached. Instead consider using `expect { }.not_to raise_error` or `expect { }.to raise_error(DifferentSpecificErrorClass)`. This message can be suppressed by setting: `RSpec::Expectations.configuration.on_potential_false_positives = :nothing`. Called from /home/marahin/Code/jsonapi_parameters/spec/lib/jsonapi_parameters/validator_spec.rb:58:in `block (5 levels) in <top (required)>'.
        does not raise validation errors
        still raises any other errors
      suppression disabled by default
WARNING: Using `expect { }.not_to raise_error(SpecificErrorClass)` risks false positives, since literally any other error would cause the expectation to pass, including those raised by Ruby (e.g. NoMethodError, NameError and ArgumentError), meaning the code you are intending to test may not even get reached. Instead consider using `expect { }.not_to raise_error` or `expect { }.to raise_error(DifferentSpecificErrorClass)`. This message can be suppressed by setting: `RSpec::Expectations.configuration.on_potential_false_positives = :nothing`. Called from /home/marahin/Code/jsonapi_parameters/spec/lib/jsonapi_parameters/validator_spec.rb:81:in `block (5 levels) in <top (required)>'.
        does not raise validation errors
```

This brings nothing but noise to the test output. Specifically, in this case, we _have_ to test for specific errors, as the `Validator` mechanism will either pass through the deeper exception, or it will raise `ValidationError` by its own. This is why we _have_ to test _specific_ error classes. 

Additionally, it seems that this false positive rule of RSpec is not smart. Consider Translator specs:

```
Translator
  with default stack limit
WARNING: Using `expect { }.not_to raise_error(SpecificErrorClass)` risks false positives, since literally any other error would cause the expectation to pass, including those raised by Ruby (e.g. NoMethodError, NameError and ArgumentError), meaning the code you are intending to test may not even get reached. Instead consider using `expect { }.not_to raise_error` or `expect { }.to raise_error(DifferentSpecificErrorClass)`. This message can be suppressed by setting: `RSpec::Expectations.configuration.on_potential_false_positives = :nothing`. Called from /home/marahin/Code/jsonapi_parameters/spec/lib/jsonapi_parameters/stack_limit_spec.rb:17:in `block (3 levels) in <top (required)>'.
    works properly if the stack level is the same as the limit
    raises an error if the stack level is above the limit
  stack limit
WARNING: Using `expect { }.not_to raise_error(SpecificErrorClass)` risks false positives, since literally any other error would cause the expectation to pass, including those raised by Ruby (e.g. NoMethodError, NameError and ArgumentError), meaning the code you are intending to test may not even get reached. Instead consider using `expect { }.not_to raise_error` or `expect { }.to raise_error(DifferentSpecificErrorClass)`. This message can be suppressed by setting: `RSpec::Expectations.configuration.on_potential_false_positives = :nothing`. Called from /home/marahin/Code/jsonapi_parameters/spec/lib/jsonapi_parameters/stack_limit_spec.rb:39:in `block (3 levels) in <top (required)>'.
    can be overwritten
WARNING: Using `expect { }.not_to raise_error(SpecificErrorClass)` risks false positives, since literally any other error would cause the expectation to pass, including those raised by Ruby (e.g. NoMethodError, NameError and ArgumentError), meaning the code you are intending to test may not even get reached. Instead consider using `expect { }.not_to raise_error` or `expect { }.to raise_error(DifferentSpecificErrorClass)`. This message can be suppressed by setting: `RSpec::Expectations.configuration.on_potential_false_positives = :nothing`. Called from /home/marahin/Code/jsonapi_parameters/spec/lib/jsonapi_parameters/stack_limit_spec.rb:48:in `block (3 levels) in <top (required)>'.
    can be overwritten using short notation
    can be reset
```

And the code within the test:
```ruby
      expect { translator.jsonapify(input) }.not_to raise_error(JsonApi::Parameters::StackLevelTooDeepError)
      expect { translator.jsonapify(input) }.not_to raise_error # To ensure this is passing
```

As you can see we are explicitly checking for a specific error, and then _ensuring_ that overall no errors were passed. This warning is completely useless and worthless for the developer and should be supressed, as it serves no value to neither the project nor the developer.

Reply:
I sign off by responding to that. 
But thanks for reply!

In this particular case (Validator) it seems to be correct but in other cases it can safe the pain.

If you set it to `RSpec::Expectations.configuration.on_potential_false_positives = :nothing`
then the problem is that you cannot disable it easily per example/line (as we can do with rubocop rules)

---------

We check only raised error on return of method, not if error was raised somewhere inside of them (we have it in other places, which does not make sense)
```ruby
  class TestedError < StandardError
  end
  class ThrownError < StandardError
  end
  class RaiseTester
    def self.throw
      raise TestedError, 'Message from test error'
    rescue TestedError => e
      raise ThrownError.new(e.message)
    end
  end

  it 'check the raise well' do
    expect { RaiseTester.throw }.to raise_error(ThrownError)
    expect { RaiseTester.throw }.to raise_error(TestedError)
  end
```
will result with:
```
1) check the raise well
     Failure/Error: expect { RaiseTester.throw }.to raise_error(TestedError)
     
       expected TestedError, got #<ThrownError: Message from test error> with backtrace:
       (...)
```

---------

Disabling false positives allow to make following mistakes:
```ruby
class TestedError < StandardError
  end
  class ThrownError < StandardError
  end
  class RaiseTester
    def self.throw
      raise TestedError, 'Message from test error'
    rescue TestedError => e
      raise ThrownError.new(e.message)
    end
  end

  it 'check the raise well' do
    expect { RaiseTester.throw }.to raise_error
  end
```
and not_to
```ruby
 class TestedError < StandardError
  end

  class RaiseTester
    def self.method(raise_error:)
      raise TestedError if raise_error == true
      'We are not using due to typo'
    end
  end

  it 'checks without being sure that we run method actually' do
    expect { RaiseTester.methoddd(raise_error: false) }.not_to raise_error(TestedError)
  end

  it 'checks with knowing that whole method did not raised anything' do
    expect { RaiseTester.methoddd(raise_error: false) }.not_to raise_error
  end
```
with 
```
WARNING: Using `expect { }.not_to raise_error(SpecificErrorClass)` risks false positives, since literally any other error would cause the expectation to pass, including those raised by Ruby (e.g. NoMethodError, NameError and ArgumentError), meaning the code you are intending to test may not even get reached. Instead consider using `expect { }.not_to raise_error` or `expect { }.to raise_error(DifferentSpecificErrorClass)`. This message can be suppressed by setting: `RSpec::Expectations.configuration.on_potential_false_positives = :nothing`.
```

---------

I would turn off this setting with :raise and disable just for Validator specs by some helper with 
```ruby
def disable_false_positives
   previous_value = RSpec::Expectations.configuration.on_potential_false_positives
   RSpec::Expectations.configuration.on_potential_false_positives = :nothing
   yield
ensure
   RSpec::Expectations.configuration.on_potential_false_positives = previous_value
end
```

Reply:
>We check only raised error on return of method, not if error was raised somewhere inside of them 

Yes. This is how validator works and this is an expected behavior. As it was discussed in the Open Source commitee tasks in the Jira and during the open discussion between the commitee members, this was our go-to strategy. 

I still stand by saying that this warning gives more noise than actual value for the developer and the library. I do not see value in displaying warnings that are not meaningful for the developers (especially since this is an expected behavior and not a bug or a misconception). 

If you would like to continue with this topic (for instance: by including the helper method you created), I would suggest joining the commitee and creating a task for this specific issue so we could discuss it in the entire matter and whether this is something we should be extra careful about or not. Until then, as it is part of the changes commitee has chosen the path for, this will not be changed in the PR.